### ISSUE TITLE
How do you get the input data when you train the network

### ISSUE THREAD

**Issue by DaisyRQin** [alandiamond/spinnaker-neuromorphic-classifier#2]
Hi. I tried to test the code. However, I found that I cannot get the input data directly from the MNIST  database images. It is said in the “A neuromorphic network for generic multivariate data classification,” that the VR responses was generated by neural gas algorithm using mdp. I am wondering how you doing that. Because if I do not know how to input data, I will never be able to test new pictures and that really bothers me. 

Could you pleased just give me some details on how to get your ClassActivation_SpikeSourceData.csv and VrResponse_SpikeSourceData.csv.

Thanks a lot.

**Comment by tnowotny** [comment]
Hi, just a quick note that Alan is on leave at the moment until in about one week's time. I am sure he'll be happy to give you some pointers then.

**Comment by DaisyRQin** [comment]
Thank you so much. I have been puzzled on this issue in about two weeks. I tried to train the neuralgasnode on mdp but still could not get the same data.

**Comment by DaisyRQin** [comment]
Hi. I am wondering if Alan is back and can give me some tips. I am quite upset these days. Thanks so much.

**Comment by alandiamond** [comment]
Hi
If you read the Method section in these 2 papers (Paper 1
http://loop.frontiersin.org/publications/44025851 , Paper 2
http://iopscience.iop.org/article/10.1088/1748-3190/11/2/026002?fromSearchPage=true)
it explains in detail how the model input is generated from the dataset
that you wish to classify.
What is it you are trying to do? Do you have a spinnaker board?
Alan

**Comment by DaisyRQin** [comment]
Thank you so much. Yes, I have a spinnaker board now, and I am trying to use this code to classify the digits pictures from MNIST database. I am doing summer intern in Imperial College right now. It is the first time I got in touch with the board.

 I still have some problems with the method, as I've read the second article you mentioned before two weeks. I am quite puzzled on how neural gas works. I installed MDP package and tried to use the neural gas to train the data. But I found nothing to output. Is that true that you initialize the neural gas Node with the num_node = 50, and then put one 28*28 image into training? Then what can I do after that?

I haven't read the first article and I will study it tonight. I really hope to get everything right. I am sorry to bother you.

Many thanks.

Rui

**Comment by alandiamond** [comment]
The MDP library's implementation of neural gas takes a numpy array as
input, where each row describes each item in the training set you have
chosen.
So for MNIST each row in the array should contain 28 contiguous sets of of
28 pixels ie. 784 data points, one for each pixel in the image.
However, the standard MNIST dataset is supplied in an unusual compressed
format which has to be decoded to create the format described above.
Its also a  large dataset and this python implementation will take a long
time to run if you gave it all 60,0000 training images.
So you need to also select a subset of images to use as your training set.

You will need to decode the raw MNIST to create the datafile you want.
There are probably python programs on the Internet that do this already, I
used a C++ program that I found.

I have attached a python program i wrote to use mdp to generate n x VRs
from a comma delimeted training data file which will be loaded into the
numpy array.
You pass the value of n and the path to the datafile as arguments as  well
as the filename in which to place the set of resulting VRs.

Having got your set of VR's you would need to create a VR response to
represent each image in numVR-dimensional  space (see papers on how the
response is worked out).
The strength of these responses determine the spike rate you use to
generate entries in the spike source file for each of the numVR input
neurons.

Its not a trivial undertaking I'm afraid ;-)

HTH

Alan

**Comment by DaisyRQin** [comment]
Thanks a lot. This is very helpful and I will try it.

Rui

**Comment by DaisyRQin** [comment]
BTW, I haven't found the code you mentioned using mdp to generate n x VRs
from a comma delimeted training data file which will be loaded into the
numpy array.

**Comment by alandiamond** [comment]
Maybe github emails removes attachments?
What is your direct email?

**Comment by DaisyRQin** [comment]
My email is daisyqin@outlook.com.

Thank you so much.

**Comment by DaisyRQin** [comment]
Hi, Alan. After running your program and read those papers, I have a question. Is it possible for me to send just one picture into the spinnaker network to classify it? 

I decode the raw MNIST data into individual pictures and transform them into rows in which contains 784 data. I used your program to create VR results, setting 50 Nodes which means 50 VRs, but found out that it won't work if I just send one row into it.  If I send two rows that will work very nice.

I am still working out how to calculate VR responses. Thanks for all the help.

**Comment by DaisyRQin** [comment]
Hi, Alan. Sorry to interrupt you again.
I used 500 images to train the neural gas node and got VRs generated by your code, which is 50 \* 784 matrix. Then I got another 500 images to calculate the VR responses. In the paper 2 http://iopscience.iop.org/article/10.1088/1748-3190/11/2/026002?fromSearchPage=true%3E there is a formula of how to calculate VR response, and I got r with 50 \* 500. But in the "A neuromorphic network for generic multivariate data classification", which is the reference paper of the original code, it has a totally different core formula. Then I got another set of r with 50 \* 500. 
However, those two r are all different from what you provided in the original code. I got each ri in [0,1], but your data is much more than that. I don't know if I need to calculate the spiking rate. But it seems in the VrResponse_SpikeSourceData.csv, you just provided VrResponses. 
And I also have some questions of the VrResponse formula. It requires maximum and minimum manhattan distance observed in the data set, or the average distance between all the input points in the set. I don't know how to do if I just have one test image. I thought I can use the training maximunm, training minimum and also training average to get the VrResponse. But I do not have your data used in the original spinnaker code. I really want to train the network by myself, but again I can't find support materials on how to get ClassActivation_SpikeSourceData.csv. I do think it contains information of image labels, without which I cannot train the network.
That's what I tried but failed again this week. I read those articles again and again but have to ask for help. I believe there's something I haven't noticed but I don't know how to do it. I am really appreciate if you can give me more tips.
Here are those two formula of calculating VrResponses,
r_temp = 1 - ( d_temp [ j ] - dmin ) / ( dmax - dmin )
or
r_temp = np.exp(-((5 \* d_temp[j])/davg)**0.7)

d_temp[j] is the distance of the jth input data and the current Vr. dmin is the minimum distance. dmax is the maximun distance. And davg is the average distance between all the input points and the current VR. 

Although my summer intern is about to finished, but I really want to know how to do these things.

Many thanks.

Rui

**Comment by alandiamond** [comment]
It sounds like you are nearly there.
You have a [0,1] response from each of 50 VR's for every entry in the test set (500 images).
You are going to present eacg image for say 200ms.
Each input neuron (50 neurons for 50 VRs) is going to rate code its corresponding VR response.
The max spike rate will be say, 200Hz represting a response of r=1.0 , 100Hz representing r=0.5 etc.
The spike source data file has a row for each input neuron (50 neurons for 50 VRs)
Each row specifies a list of times (ms) that the neuron should spike.
The times depend on the spike frequency required.
Imagine that VR0 responds with r=0.5 for the first image and r=1 for the second image.
100Hz implies spike times at 10ms intervals (0,10,20,30...200)
After 200ms we switch to spikes at 200Hz, implies spike times at 5ms intervals (205,210,215,220...400)
So in the first row of the spike source data we would see
0,10,20,30...200, 205,210,215,220...400

THis is what you see in the example spike source file.
The only difference is that to avoid every neuron spiking at t=0 the first spike time is dithered, (eg selected randmly between 0 and 10).

The ClassActivation_SpikeSourceData.csv is the same format, but targets the 10 output neurons (one per class lable) during training to create spiking activity in the correct output . This causes a Hebbian association (weight potentiation) to be created by STDP between PN neurons and a particular class.
You should see in the spike source data that each row has large gaps bewtween firing periods, this is becuase each row represents a single class (digit) and only appears in the data every 10th digit.

**Comment by DaisyRQin** [comment]
Thank you so much, Alan. It sounds very interesting. I will try what I can do tonight.
